project_recalculate/models/project_task.py

```diff
@@ -0,0 +1,120 @@
+        days = (date_end - date_start).days
+        return sum(1 for x in xrange(days)
+                   if (date_start + timedelta(x)).weekday() < 5)
+
+    def count_days_weekend(self, date_start, date_end):
+        days = (date_end - date_start).days
+        return sum(1 for x in xrange(days)
+                   if (date_start + timedelta(x)).weekday() >= 5)
+
+    def correct_days_to_workable(self, date, increment=True):
+        while date.weekday() >= 5:
+            if increment:
+                date += timedelta(days=1)
+            else:
+                date -= timedelta(days=1)
+        return date
+
+    def calculate_date_without_weekend(self, date_start, days, increment=True):
+        total_days = 0
+        c = 0
+        while days != total_days:
+            if total_days > 0:
+                recalculate = days + (days - total_days) + c
+                c += 1
+            else:
+                recalculate = days
+            if increment:
+                start = date_start
+                end = date_start + timedelta(days=recalculate)
+            else:
+                start = date_start - timedelta(days=recalculate)
+                end = date_start
+            total_days = self.count_days_without_weekend(start, end)
+        if total_days > 0:
+            date = end if increment else start
+        else:
+            date = date_start
+        return self.correct_days_to_workable(date, increment)
+
+    def on_change_dates(self, date_start, date_end, vals):
+        vals['estimated_days'] = self.count_days_without_weekend(
+            fields.Datetime.from_string(date_start),
+            fields.Datetime.from_string(date_end))
+        calculation_type = self.project_id.calculation_type
+        if calculation_type:
+            if calculation_type == 'date_begin':
+                aux_date_end = date_start
+                aux_date_start = self.project_id.date_start
+            else:
+                aux_date_end = self.project_id.date
+                aux_date_start = date_start
+            vals['from_days'] = self.count_days_without_weekend(
+                fields.Datetime.from_string(aux_date_start),
+                fields.Datetime.from_string(aux_date_end))
+        return vals
+
+    @api.multi
+    def write(self, vals):
+        if (not self.env.context.get('project_recalculate')
+                and (vals.get('date_start') or vals.get('date_end'))):
+            date_start = (vals.get('date_start')
+                          if vals.get('date_start') else self.date_start)
+            date_end = (vals.get('date_end')
+                        if vals.get('date_end') else self.date_end)
+            vals = self.on_change_dates(date_start, date_end, vals)
+        return super(ProjectTask, self).write(vals)
+
+    def task_recalculate(self):
+        self.ensure_one()
+        increment = (True if self.project_id.calculation_type == 'date_begin'
```

Comment:
This should be enough:

        increment = self.project_id.calculation_type == 'date_begin'